File: examples/AdWords/CSharp/v201603/CampaignManagement/PromoteExperiment.cs
// Copyright 2016, Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using Google.Api.Ads.AdWords.Lib;
using Google.Api.Ads.AdWords.v201603;

using System;
using System.Collections.Generic;
using System.IO;

namespace Google.Api.Ads.AdWords.Examples.CSharp.v201603 {
  /// <summary>
  /// This example promotes an experiment, which permanently applies all the
  /// experiment changes made to its related ad groups, criteria and ads. To
  /// create an experiment, run AddExperiment.vb.
  /// </summary>
  public class PromoteExperiment : ExampleBase {
    /// <summary>
    /// Main method, to run this code example as a standalone application.
    /// </summary>
    /// <param name="args">The command line arguments.</param>
    public static void Main(string[] args) {
      PromoteExperiment codeExample = new PromoteExperiment();
      Console.WriteLine(codeExample.Description);
      try {
        long experimentId = long.Parse("INSERT_EXPERIMENT_ID_HERE");
        codeExample.Run(new AdWordsUser(), experimentId);
      } catch (Exception e) {
        Console.WriteLine("An exception occurred while running this code example. {0}",
            ExampleUtilities.FormatException(e));
      }
    }

    /// <summary>
    /// Returns a description about the code example.
    /// </summary>
    public override string Description {
      get {
        return "This example promotes an experiment, which permanently applies all the experiment" +
            " changes made to its related ad groups, criteria and ads. To create an experiment, " +
            "run AddExperiment.cs.";
      }
    }

    /// <summary>
    /// Runs the code example.
    /// </summary>
    /// <param name="user">The AdWords user.</param>
    /// <param name="experimentId">Id of the experiment to be promoted.</param>
    public void Run(AdWordsUser user, long experimentId) {
      // Get the ExperimentService.
      ExperimentService experimentService =
          (ExperimentService) user.GetService(AdWordsService.v201603.ExperimentService);

      // Set experiment's status to PROMOTED.
      Experiment experiment = new Experiment();
      experiment.id = experimentId;
      experiment.status = ExperimentStatus.PROMOTED;

      // Create the operation.
      ExperimentOperation operation = new ExperimentOperation();
      operation.@operator = Operator.SET;
      operation.operand = experiment;

      try {
        // Update the experiment.
        ExperimentReturnValue retVal = experimentService.mutate(
            new ExperimentOperation[] {operation});

        // Display the results.
        if (retVal != null && retVal.value != null && retVal.value.Length > 0) {
          Experiment promotedExperiment = retVal.value[0];
          Console.WriteLine("Experiment with name = \"{0}\" and id = \"{1}\" was promoted.\n",
              promotedExperiment.name, promotedExperiment.id);
        } else {
          Console.WriteLine("No experiments were promoted.");
        }
      } catch (Exception e) {
        throw new System.ApplicationException("Failed to promote experiment.", e);
      }
    }
  }
}

File: examples/AdWords/Vb/v201603/CampaignManagement/PromoteExperiment.vb
' Copyright 2016, Google Inc. All Rights Reserved.
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
'     http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.

Imports Google.Api.Ads.AdWords.Lib
Imports Google.Api.Ads.AdWords.v201603

Imports System
Imports System.Collections.Generic
Imports System.IO

Namespace Google.Api.Ads.AdWords.Examples.VB.v201603
  ''' <summary>
  ''' This example promotes an experiment, which permanently applies all the
  ''' experiment changes made to its related ad groups, criteria and ads. To
  ''' create an experiment, run AddExperiment.vb.
  ''' </summary>
  Public Class PromoteExperiment
    Inherits ExampleBase
    ''' <summary>
    ''' Main method, to run this code example as a standalone application.
    ''' </summary>
    ''' <param name="args">The command line arguments.</param>
    Public Shared Sub Main(ByVal args As String())
      Dim codeExample As New PromoteExperiment
      Console.WriteLine(codeExample.Description)
      Try
        Dim experimentId As Long = Long.Parse("INSERT_EXPERIMENT_ID_HERE")
        codeExample.Run(New AdWordsUser, experimentId)
      Catch e As Exception
        Console.WriteLine("An exception occurred while running this code example. {0}", _
            ExampleUtilities.FormatException(e))
      End Try
    End Sub

    ''' <summary>
    ''' Returns a description about the code example.
    ''' </summary>
    Public Overrides ReadOnly Property Description() As String
      Get
        Return "This example promotes an experiment, which permanently applies all the " & _
            "experiment changes made to its related ad groups, criteria and ads. To create an " & _
            "experiment, run AddExperiment.vb."
      End Get
    End Property

    ''' <summary>
    ''' Runs the code example.
    ''' </summary>
    ''' <param name="user">The AdWords user.</param>
    ''' <param name="experimentId">Id of the experiment to be promoted.</param>
    Public Sub Run(ByVal user As AdWordsUser, ByVal experimentId As Long)
      ' Get the ExperimentService.
      Dim experimentService As ExperimentService = CType(user.GetService( _
          AdWordsService.v201603.ExperimentService), ExperimentService)

      ' Set experiment's status to PROMOTED.
      Dim experiment As New Experiment
      experiment.id = experimentId
      experiment.status = ExperimentStatus.PROMOTED

      ' Create the operation.
      Dim operation As New ExperimentOperation
      operation.operator = [Operator].SET
      operation.operand = experiment

      Try
        ' Update the experiment.
        Dim retVal As ExperimentReturnValue = experimentService.mutate( _
            New ExperimentOperation() {operation})

        ' Display the results.
        If ((Not retVal Is Nothing) AndAlso (Not retVal.value Is Nothing) AndAlso _
            (retVal.value.Length > 0)) Then
          Dim promotedExperiment As Experiment = retVal.value(0)
          Console.WriteLine("Experiment with name = ""{0}"" and id = ""{1}"" was promoted.", _
                promotedExperiment.name, promotedExperiment.id)
        Else
          Console.WriteLine("No experiments were promoted.")
        End If
      Catch e As Exception
        Throw New System.ApplicationException("Failed to promote experiment(s).", e)
      End Try
    End Sub
  End Class
End Namespace
